bdb/llmeta.c

```diff
@@ -4298,6 +4301,93 @@ int bdb_set_schema_change_history(tran_type *t, const char *tablename,
     return rc;
 }
 
+void bdb_llmeta_free_schema_versions(schema_version_row *data, int n)
+{
+    for (int i = 0; i < n; i++) {
+        if (data[i].db_name != NULL) {
+            free(data[i].db_name);
+        }
+        if (data[i].csc2 != NULL) {
+            free(data[i].csc2);
+        }
+    }
+    free(data);
+}
+
+int bdb_llmeta_get_schema_versions(tran_type *t, schema_version_row **data, int *num,
+                                    int *bdberr)
+{
+    int len, version, rowidx;
+    int rc = 0, dbnums[MAX_NUM_TABLES], fndnumtbls;
+    char *tblnames[MAX_NUM_TABLES];
+    char *db_name;
+
+    *data = NULL;
+    *num = 0;
+    rowidx = 0;
+
+    rc = bdb_llmeta_get_tables(t, tblnames, dbnums, MAX_NUM_TABLES, &fndnumtbls, bdberr);
+    if (rc || *bdberr != BDBERR_NOERROR) {
+        logmsg(LOGMSG_ERROR, "%s: Failed to get tables from llmeta\n", __func__);
+        rc = rc ? rc : 1;
+        goto err;
+    }
+
+    for (int tableidx = 0; tableidx < fndnumtbls; tableidx++) {
+        db_name = tblnames[tableidx];
+        version = get_csc2_version_tran(db_name, t);
+        if (version == -1) {
+            logmsg(LOGMSG_ERROR, "%s: Could not find csc2 version for table %s\n", __func__, db_name);
+            rc = 1;
+            goto err;
+        }
+        *num += version;
+    }
+
+    *data = calloc(*num, sizeof(schema_version_row));
+
+    for (int tableidx = 0; tableidx < fndnumtbls; tableidx++) {
+        db_name = tblnames[tableidx];
+        
+        version = get_csc2_version_tran(db_name, t);
+        if (version == -1) {
+            logmsg(LOGMSG_ERROR, "%s: Could not find csc2 version for table %s\n", __func__, db_name);
+            rc = 1;
+            goto err;
+        }
+
+        for (int vers=1; vers<=version; vers++) {
+            if (rowidx >= *num) {
+                logmsg(LOGMSG_ERROR, "%s: More versions found than anticipated.\n", __func__);
+                rc = 1;
+                goto err;
+            }
+            schema_version_row *row = *data + rowidx++;
+
+            row->db_name = strdup(db_name);
+            row->vers = vers;
+            rc = get_csc2_file_tran(db_name, vers, &row->csc2, &len, t);
+            if (rc) {
+                logmsg(LOGMSG_ERROR, "%s: Failed to get csc2 file for db %s and version %d\n",
+                        __func__, db_name, vers);
+                goto err;
+            }
+        }
+    }
+
+err:
+
+    if (rc != 0) {
+       if (*data != NULL) {
+            bdb_llmeta_free_schema_versions(*data, rowidx);
+            *data = NULL;
+       }
+       *num = 0;
+    }
+
+    return rc;
+}
```

Comment:
Changes to llmeta cannot occur while this function runs or it will produce undefined behavior. Other threads modifying llmeta are excluded by [acquiring the `comdb2_tables` lock over systable access](https://github.com/bloomberg/comdb2/pull/4191/files#diff-5c0dcb2e559d2ca33fe51f91eaba821f0162e155618915f0f36cee59aaa5951bR67).